### PR TITLE
IGNITE-16185 Improve the interface of UserObjectMarshaller

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-16185

1. Now unmarshal() method receives a registry of remote descriptors
2. It builds a context from them
3. The context is being passed through whole unmarshalling process (this caused most of the changes)
4. Serialization special methods are now cached in descriptors themselves, so no lifetime problem occurs anymore between local descirptors (that live forever) and remote descriptors (that are discarded when their session is closed).

### CODE REVIEW

**Review by SammyVimes — CHANGES_REQUESTED** [review]

File: modules/network/src/main/java/org/apache/ignite/internal/network/serialization/ClassDescriptor.java [hunk 1]
```diff
@@ -211,6 +215,37 @@ public boolean supportsWriteReplace() {
...
     }
 
+    /**
+     * Applies writeReplace() method.
+     *
+     * @param object     object to which to apply
+     * @return writeReplace() result
+     * @throws SpecialMethodInvocationException if writeReplace() invocation fails
+     */
+    public Object applyWriteReplace(Object object) throws SpecialMethodInvocationException {
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/serialization/ClassDescriptor.java]: Moved this functionality out of `ClassDescriptor`; instead, `ClassDescriptor` returns an instance of `SpecialSerializationMethods` that can be used to actually apply the methods.

**Review by SammyVimes — APPROVED** [review]
LGTM

**Comment by SammyVimes** [comment]
Merged to the main branch https://github.com/apache/ignite-3/commit/82d2e23096571bc640677fb1225fb72955965530

**Comment by rpuch** [comment]
Thanks for the review

**Pull request closed without merging**